lifecycle.go

```diff
@@ -0,0 +1,88 @@
+package testcontainers
+
+import (
+	"context"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/go-connections/nat"
+)
+
+func (p *DockerProvider) preCreateContainerHook(ctx context.Context, req ContainerRequest, dockerInput *container.Config, hostConfig *container.HostConfig, networkingConfig *network.NetworkingConfig) error {
+	// prepare mounts
+	hostConfig.Mounts = mapToDockerMounts(req.Mounts)
+
+	endpointSettings := map[string]*network.EndpointSettings{}
+
+	// #248: Docker allows only one network to be specified during container creation
+	// If there is more than one network specified in the request container should be attached to them
+	// once it is created. We will take a first network if any specified in the request and use it to create container
+	if len(req.Networks) > 0 {
+		attachContainerTo := req.Networks[0]
+
+		nw, err := p.GetNetwork(ctx, NetworkRequest{
+			Name: attachContainerTo,
+		})
+		if err == nil {
+			aliases := []string{}
+			if _, ok := req.NetworkAliases[attachContainerTo]; ok {
+				aliases = req.NetworkAliases[attachContainerTo]
+			}
+			endpointSetting := network.EndpointSettings{
+				Aliases:   aliases,
+				NetworkID: nw.ID,
+			}
+			endpointSettings[attachContainerTo] = &endpointSetting
+		}
+	}
+
+	if req.ConfigModifier != nil {
+		req.ConfigModifier(dockerInput)
+	}
+
+	if req.HostConfigModifier == nil {
+		req.HostConfigModifier = defaultHostConfigModifier(req)
+	}
+	req.HostConfigModifier(hostConfig)
+
+	if req.EnpointSettingsModifier != nil {
+		req.EnpointSettingsModifier(endpointSettings)
+	}
```

Comment:
Shouldn't the callbacks be called after the mapping has been completed? For example, lines 71 and 72 modify the instance again, if I have understood it correctly.

Reply:
L54:72 are sanitisations that are needed in the case there are no exposed ports, where we read the raw image and extract the ports from there if possible.

Reply:
Yes, but that means that the modifier methods do not receive the final state of the request.

Reply:
That's correct: we could probably move the modifiers after that 🤔

Reply:
I've updated it to run the modifiers last, sanitising right before them

Reply:
Mmm not sure if I should step back on this change: in the case the hostConfig modifier does not update the exposed ports, neither the original container request does it and the container runs in container network, the original code checked the image ports and exported them. I think we should keep that behaviour and document it

Reply:
Okay, that is the only thing I noticed. However, please keep in mind that my skills in Go are limited 🤔. Therefore, I was not able to comprehend all the consequences. .NET passes the final [instance](https://github.com/testcontainers/testcontainers-dotnet/blob/fe9f55a84341b57f5e054e631043f590dd2cc74c/src/Testcontainers/Clients/DockerContainerOperations.cs#L173-L208) to the modifier.